### PR TITLE
feat: expose if docker is enabled and wait for docker to be ready

### PR DESCRIPTION
**Pull request by toast-gear**
Ref https://github.com/actions-runner-controller/actions-runner-controller/issues/897 https://github.com/actions-runner-controller/actions-runner-controller/pull/915

Having M1 architecture issues when testing but the controller change can be merged prior. 

The environment variable punches through now and should be referenceable in the entrypoint of the runner

Docker enabled:

image

Docker disabled:

image

### CODE REVIEW

File: runner/Makefile [hunk 1]
```diff
@@ -1,9 +1,10 @@
-NAME ?= summerwind/actions-runner
-DIND_RUNNER_NAME ?= ${NAME}-dind
+DOCKER_USER ?= summerwind
```
Comment (awoimbee) [inline runner/Makefile:1]: I don't think this is needed

File: runner/entrypoint.sh [hunk 2]
```diff
@@ -24,6 +24,16 @@ if [ ! -z "${STARTUP_DELAY_IN_SECONDS}" ]; then
   sleep ${STARTUP_DELAY_IN_SECONDS}
 fi
 
+if [ "${DOCKER_ENABLED}" == "true" ] && [ "${DISABLE_WAIT_FOR_DOCKER}" != "true" ]; then
+  log "Docker enabled runner detected and Docker daemon wait is enabled"
```
Comment (awoimbee) [inline runner/entrypoint.sh:28]:
```suggestion
  log "Docker is enabled and Docker daemon wait is enabled"
```

File: controllers/runner_controller.go [hunk 3]
```diff
@@ -775,6 +775,10 @@ func newRunnerPod(template corev1.Pod, runnerSpec v1alpha1.RunnerConfig, default
 			Name:  "RUNNER_GROUP",
 			Value: runnerSpec.Group,
 		},
+		{
+			Name:  "DOCKER_ENABLED",
```
Comment (awoimbee) [inline controllers/runner_controller.go:779]: You are not passing `DISABLE_WAIT_FOR_DOCKER` to the runner ? I don't think this PR works ?

File: runner/Makefile [hunk 4]
```diff
@@ -1,9 +1,10 @@
...
+RUNNER_NAME ?= ${DOCKER_USER}/actions-runner
+DIND_RUNNER_NAME ?= ${DOCKER_USER}/actions-runner-dind
 TAG ?= latest
-TARGET_PLATFORM ?= $(shell arch)
+TARGETPLATFORM ?= $(shell arch)
```
Comment (awoimbee) [inline runner/Makefile:5]: Would've been better to move from `TARGETPLATFORM` -> `TARGET_PLATFORM` than the inverse...

**Review by awoimbee — COMMENTED** [review]
> Having M1 architecture issues when testing but the controller change can be merged prior.

So... these changes are untested but "can be merged prior"

File: runner/Makefile [hunk 5]
```diff
@@ -1,9 +1,10 @@
-NAME ?= summerwind/actions-runner
-DIND_RUNNER_NAME ?= ${NAME}-dind
+DOCKER_USER ?= summerwind
+RUNNER_NAME ?= ${DOCKER_USER}/actions-runner
+DIND_RUNNER_NAME ?= ${DOCKER_USER}/actions-runner-dind
```
Comment (awoimbee) [inline runner/Makefile:3]: This breaks the instructions in CONTRIBUTING.md
> - Run `NAME=$DOCKER_USER/actions-runner-controller VERSION=canary make docker-build docker-push` for a custom container image build

**Comment by toast-gear** [comment]
> > Having M1 architecture issues when testing but the controller change can be merged prior.
> 
> So... these changes are untested but "can be merged prior"

My apologies, I should have updated the PR really to reflect reality. Spoken to @mumoshu offline about this so I hadnt bothered to update the PR etc.

**Comment by toast-gear** [comment]
> > Having M1 architecture issues when testing but the controller change can be merged prior.
> 
> So... these changes are untested but "can be merged prior"

My apologies, I should have updated the PR really to reflect reality. Spoken to @mumoshu offline about this so I hadnt bothered to update the PR etc.

File: runner/Makefile (hunk 5 above)
Comment (toast-gear) [reply runner/Makefile:3]: The instructions within the `CONTRIBUTING.md` refer to the Makefile within the root of the repo, these changes are within `runner/`Makefile`

File: runner/entrypoint.sh (hunk 2 above)
Comment (toast-gear) [reply runner/entrypoint.sh:28]: This could be worded better tbf, I'll make the change

File: controllers/runner_controller.go (hunk 3 above)
Comment (toast-gear) [reply controllers/runner_controller.go:779]: PR was poorly named

File: runner/Makefile (hunk 4 above)
Comment (toast-gear) [reply runner/Makefile:5]: Docker convetion is upper case no spaces so I followed that tbh e.g. `TARGETPLATFORM`, `BUILDPLATFORM`, `TARGETVARIANT` etc etc

**Comment by awoimbee** [comment]
> My apologies

Not problem, thanks for opening this PR.

But it needs a good cleaning ! I didn't know how fast you would respond so I worked on it a bit (rebased, squashed, improved): 
https://github.com/actions-runner-controller/actions-runner-controller/compare/master...Extrality:runner-wait-for-dind

File: runner/Makefile [hunk 6]
```diff
@@ -1,9 +1,10 @@
-NAME ?= summerwind/actions-runner
-DIND_RUNNER_NAME ?= ${NAME}-dind
+DOCKER_USER ?= summerwind
+RUNNER_NAME ?= ${DOCKER_USER}/actions-runner
```
Comment (mumoshu) [inline runner/Makefile:2]: We'd need to update this part of your contribution guide as well

CleanShot 2021-12-12 at 15 22 37@2x

File: controllers/runner_controller.go [hunk 7]
```diff
@@ -775,6 +775,10 @@ func newRunnerPod(template corev1.Pod, runnerSpec v1alpha1.RunnerConfig, default
...
+			Value: fmt.Sprintf("%v", dockerEnabled),
```
Comment (mumoshu) [inline controllers/runner_controller.go:780]: AFAIK `dockerEnabled` doesn't turn `true` for `dockerdWithinRunnerContainer` where you'd also want it to wait for dockerd startup.

How about:

```suggestion
			Value: fmt.Sprintf("%v", dockerEnabled || dockerdInRunner),
```

File: runner/Makefile (hunk 6 above)
Comment (mumoshu) [reply runner/Makefile:2]: Addressed in aab3fae

**Review by mumoshu — APPROVED** [review]
LGTM. Thanks a lot for your efforts @toast-gear!

Credits also go to @artificial-aidan for submitting #915 which provided the foundational work that resulted in this change, and everyone provided various review comments and feedbacks. Thanks! ☺️

**Comment by genisd** [comment]
ha very nice! :heart: 
merging this into our branch right away, go be gone race condition!

**Pull request merged**